rpc/swap.go

```diff
@@ -303,6 +303,10 @@ func (s *SwapService) Cancel(_ *http.Request, req *CancelRequest, resp *CancelRe
 		ss = s.xmrmaker.GetOngoingSwapState(req.OfferID)
 	}
 
+	if ss == nil {
+		return fmt.Errorf("failed to find swap state with ID %s", req.OfferID)
+	}
```

Comment:
this was an attempt to fix a panic in `cancel()` when it's called twice successively:
```
2023-03-26T00:40:48.371Z	INFO	xmrmaker	xmrmaker/event.go:241	EventExit
2023-03-26T00:40:48.371Z	DEBUG	xmrmaker	xmrmaker/swap_state.go:338	attempting to exit swap: nextExpectedEvent=EventETHLockedType
2023-03-26T00:40:48.371Z	DEBUG	rpc	rpc/ws.go:61	failed to read websockets message: websocket: close 1006 (abnormal closure): unexpected EOF
2023-03-26T00:40:48.372Z	INFO	xmrmaker	xmrmaker/swap_state.go:347	exit status Aborted
2023-03-26T00:40:48.373Z	DEBUG	xmrmaker	xmrmaker/swap_state.go:356	re-added offer 0xadc8490398ecf78797aebd055ca13f698652340faf8de05367f806bdb26e634b
2023-03-26T00:40:48.377Z	INFO	xmrmaker	xmrmaker/swap_state.go:383	**swap aborted: id=0xadc8490398ecf78797aebd055ca13f698652340faf8de05367f806bdb26e634b**
2023-03-26T00:40:48.377Z	DEBUG	net	net/host.go:179	closing stream: peer=12D3KooWDqCzbjexHEa8Rut7bzxHFpRMZyDRW1L6TGkL1KY24JH5 protocol=/atomic-swap/0.2/5/swap/0
2023-03-26T00:40:48.377Z	DEBUG	net	net/initiate.go:149	Failed to read message from peer, id=12D3KooWDq-55-86 protocol=/atomic-swap/0.2/5/swap/0: stream reset
2023-03-26T00:40:48.377Z	DEBUG	net	net/initiate.go:131	closing stream: peer=12D3KooWDqCzbjexHEa8Rut7bzxHFpRMZyDRW1L6TGkL1KY24JH5 protocol=/atomic-swap/0.2/5/swap/0
2023-03-26T00:40:48.377Z	DEBUG	net	net/initiate.go:134	exiting swap...
2023/03/26 00:42:34 http: panic serving 127.0.0.1:38580: runtime error: invalid memory address or nil pointer dereference
goroutine 3013 [running]:
net/http.(*conn).serve.func1()
	/usr/local/go/src/net/http/server.go:1850 +0xbf
panic({0x122e320, 0x2114cd0})
	/usr/local/go/src/runtime/panic.go:890 +0x262
github.com/athanorlabs/atomic-swap/protocol/xmrmaker.(*swapState).Exit(0x0)
	/root/atomic-swap/protocol/xmrmaker/swap_state.go:332 +0x77
github.com/athanorlabs/atomic-swap/rpc.(*SwapService).Cancel(0xc0077cf6e0, 0xc0073abc68?, 0xc0071dac80, 0xc000417d38)
	/root/atomic-swap/rpc/swap.go:306 +0x262
reflect.Value.call({0xc0077d1a80?, 0xc0076b9c20?, 0xc000881590?}, {0x13f5267, 0x4}, {0xc000881740, 0x4, 0xc0008815c0?})
	/usr/local/go/src/reflect/value.go:584 +0x8c5
reflect.Value.Call({0xc0077d1a80?, 0xc0076b9c20?, 0xc0071dac80?}, {0xc000881740?, 0x203000?, 0x203000?})
	/usr/local/go/src/reflect/value.go:368 +0xbc
github.com/gorilla/rpc/v2.(*Server).ServeHTTP(0xc0077eb180, {0x1872750, 0xc0082bfce0}, 0xc0006a2500)
	/root/go/pkg/mod/github.com/gorilla/rpc@v1.2.0/v2/server.go:220 +0x7f7
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00773ad80, {0x1872750, 0xc0082bfce0}, 0xc0006a2300)
	/root/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
github.com/gorilla/handlers.(*cors).ServeHTTP(0xc0077990e0, {0x1872750, 0xc0082bfce0}, 0xc0006a2300)
	/root/go/pkg/mod/github.com/gorilla/handlers@v1.5.1/cors.go:54 +0x370
net/http.serverHandler.ServeHTTP({0x186e910?}, {0x1872750, 0xc0082bfce0}, 0xc0006a2300)
	/usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc0072086e0, {0x1873be0, 0xc0077fe660})
	/usr/local/go/src/net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3102 +0x4db
```

I think a lock somewhere could also fix this.

Reply:
What happens if two `Cancel(...)` requests come in at the same time? They will both succeed in getting the same `ss` variable and then go on to call `Exit` on that same object simultaneously. Is that code re-entryancy safe? If it is, just add some code comments to that effect. If the code is not re-entrant, we'll need locking.

Reply:
I see one more problem.  `GetOngoingSwapState` in the `xmrtaker` isn't grabbing the lock where it should. In `xmrmaker` it is.

Reply:
`Exit()` is safe for concurrent calls, since it puts an exit event into the eventCh, and events are handled sequentially

however the `Refund` and `Cancel` methods could probably be tested more. we should consider removing `Refund` in favour of `Cancel`

Reply:
Can you create that ticket? I could create it, but I'm not sure I can describe the finer nuances. If Cancel can already do everything that Refund does, we should definitely remove Refund and just adjust the documentation.